api/src/main/java/org/apache/iceberg/PartitionSpec.java

```diff
@@ -346,11 +346,34 @@ private void checkAndAddPartitionName(String name, Integer identitySourceColumnI
       partitionNames.add(name);
     }
 
+    private String getDedupKey(PartitionField field) {
+      String transform = field.transform().toString();
+      String type;
+      if ("hour".equalsIgnoreCase(transform) || "day".equalsIgnoreCase(transform) ||
+          "month".equalsIgnoreCase(transform) || "year".equalsIgnoreCase(transform)) {
+        type = "time";
+      } else if (transform.startsWith("bucket[")) {
+        type = "bucket";
+      } else {
+        return null;
```

Comment:
What about using the transform name in this situation? Then this would always catch duplicate transforms.

Also, should we add a case for `truncate` with different lengths?